ratbagd/ratbagd.c

```diff
@@ -181,6 +181,148 @@ static int ratbagd_get_themes(sd_bus *bus,
+	.num_resolutions = 3,
+	.num_buttons = 4,
+	.num_leds = 3,
+	.profiles = {
+		{
+			.buttons = {
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_BUTTON,
+				  .button = 0 },
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_KEY,
+				  .key = 4 },
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_SPECIAL,
+				  .special = RATBAG_BUTTON_ACTION_SPECIAL_PROFILE_CYCLE_UP },
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_MACRO },
+			},
+			.resolutions = {
+				{ .xres = 100, .yres = 200, .hz = 1000 },
+				{ .xres = 200, .yres = 300, .hz = 1000, .active=true, .dflt=true },
+				{ .xres = 300, .yres = 400, .hz = 1000 },
+			},
+			.active = true,
+			.dflt = false,
+			.leds = {
+				{
+					.mode = RATBAG_LED_OFF,
+					.color = { .red = 255, .green = 0,	.blue = 0 },
+					.hz = 1,
+					.brightness = 20
+				},
+				{
+					.mode = RATBAG_LED_ON,
+					.color = { .red = 255, .green = 0,	.blue = 0 },
+					.hz = 1,
+					.brightness = 20
+				},
+				{
+					.mode = RATBAG_LED_CYCLE,
+					.color = { .red = 255, .green = 255, .blue = 0 },
+					.hz = 3,
+					.brightness = 40
+				}
+			},
+		},
+		{
+			.buttons = {
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_KEY,
+				  .key = 4 },
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_KEY,
+				  .key = 5 },
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_KEY,
+				  .key = 6 },
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_KEY,
+				  .key = 7 },
+			},
+			.resolutions = {
+				{ .xres = 1100, .yres = 1200, .hz = 2000 },
+				{ .xres = 1200, .yres = 1300, .hz = 2000, .dflt=true },
+				{ .xres = 1300, .yres = 1400, .hz = 2000, .active=true },
+			},
+			.active = false,
+			.dflt = true,
+		},
+		{
+			.buttons = {
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_SPECIAL,
+				  .special = RATBAG_BUTTON_ACTION_SPECIAL_PROFILE_CYCLE_UP },
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_MACRO,
+				  .button = 1 },
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_BUTTON,
+				  .button = 2 },
+				{ .type = RATBAG_BUTTON_ACTION_TYPE_BUTTON,
+				  .button = 3 },
+			},
+			.resolutions = {
+				{ .xres = 2100, .yres = 2200, .hz = 3000, .active=true },
+				{ .xres = 2200, .yres = 2300, .hz = 3000, .dflt=true },
+				{ .xres = 2300, .yres = 2400, .hz = 3000 },
+			},
+			.leds = {
+				{
+					.mode = RATBAG_LED_ON,
+					.color = { .red = 255, .green = 0,	.blue = 0 },
+					.hz = 1,
+					.brightness = 20
+				},
+				{
+					.mode = RATBAG_LED_CYCLE,
+					.color = { .red = 255, .green = 255, .blue = 0 },
+					.hz = 3,
+					.brightness = 40
+				}
+		},
+		.active = false,
+		.dflt = false,
+		},
+	},
+	.destroyed = NULL,
+	.destroyed_data = NULL,
+};
+
+static int ratabgd_create_test_device(struct ratbagd *ctx)
+{
+	struct ratbagd_device *ratbagd_test_device = NULL;
+	struct ratbag_device *device;
+	int r;
+
+	device = ratbag_device_new_test_device(ctx->lib_ctx, &ratabgd_sane_device);
+
+	r = ratbagd_device_new(&ratbagd_test_device, ctx, "test_device", device);
+
+	/* the ratbagd_device takes its own reference, drop ours */
+	ratbag_device_unref(device);
```

Comment:
Well, no. And actually, the error path is wrong in ratbagd_device_new().

Theorically, ratbagd_device_new() calls ratbag_device_ref() on the ratbag device and should unref it in case of an error. We do not want to keep the reference around anyway, so we should unref our own ratbag_device here so that if r is negative or null, both cases are fine.

The bug is more likely in ratbagd_device_new() which doesn't unref the device in case of a failure (though the failures cases are unlikely to happen on a normal execution path).